File: behavioral-pattern/template.js
var __extends = (this && this.__extends) || (function () {
    var extendStatics = function (d, b) {
        extendStatics = Object.setPrototypeOf ||
            ({ __proto__: [] } instanceof Array && function (d, b) { d.__proto__ = b; }) ||
            function (d, b) { for (var p in b) if (Object.prototype.hasOwnProperty.call(b, p)) d[p] = b[p]; };
        return extendStatics(d, b);
    };
    return function (d, b) {
        if (typeof b !== "function" && b !== null)
            throw new TypeError("Class extends value " + String(b) + " is not a constructor or null");
        extendStatics(d, b);
        function __() { this.constructor = d; }
        d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
    };
})();
var AbstractClass = /** @class */ (function () {
    function AbstractClass() {
    }
    AbstractClass.prototype.templateMethod = function () {
        this.baseOperation1();
        this.requiredOperation1();
        this.baseOperation2();
        this.hook1();
        this.requiredOperation2();
        this.baseOperation3();
        this.hook2();
    };
    AbstractClass.prototype.baseOperation1 = function () {
        console.log("AbstractClass says: I am doing the bulk of the work");
    };
    AbstractClass.prototype.baseOperation2 = function () {
        console.log("AbstractClass says: But I let subclass override some operations");
    };
    AbstractClass.prototype.baseOperation3 = function () {
        console.log("AbstractClass says: but I am doing the bulk of the work anyway");
    };
    AbstractClass.prototype.hook1 = function () { };
    AbstractClass.prototype.hook2 = function () { };
    return AbstractClass;
}());
var ConcreteClass1 = /** @class */ (function (_super) {
    __extends(ConcreteClass1, _super);
    function ConcreteClass1() {
        return _super !== null && _super.apply(this, arguments) || this;
    }
    ConcreteClass1.prototype.requiredOperation1 = function () {
        console.log("ConcreteClass1 says: Implemented Operation1");
    };
    ConcreteClass1.prototype.requiredOperation2 = function () {
        console.log("ConcreteClass2 says: Implemented Operation2");
    };
    return ConcreteClass1;
}(AbstractClass));
var ConcreteClass2 = /** @class */ (function (_super) {
    __extends(ConcreteClass2, _super);
    function ConcreteClass2() {
        return _super !== null && _super.apply(this, arguments) || this;
    }
    ConcreteClass2.prototype.requiredOperation1 = function () {
        console.log("ConcreteClass2 says: Implemented Operation1");
    };
    ConcreteClass2.prototype.requiredOperation2 = function () {
        console.log("ConcreteClass2 says: Implemented Operation2");
    };
    ConcreteClass2.prototype.hook1 = function () {
        console.log("ConcreteClass2 says: Overridden Hook1");
    };
    return ConcreteClass2;
}(AbstractClass));
function clientCode(abstract) {
    abstract.templateMethod();
}
console.log("Same client code can work with different subclass: ");
clientCode(new ConcreteClass1());
console.log("");
console.log("Same client code can work with different subclass: ");
clientCode(new ConcreteClass2());

File: behavioral-pattern/template.py


from abc import ABC, abstractmethod


class AbstractClass(ABC):

    def template_method(self) -> None:
        self.base_operation1()
        self.base_operation2()
        self.base_operation3()
        self.required_operation1()
        self.required_operation2()
        self.hook1()
        self.hook2()

    def base_operation1(self) -> None:
        pass

    def base_operation2(self) -> None:
        pass

    def base_operation3(self) -> None:
        pass

    @abstractmethod
    def required_operation1(self) -> None:
        pass

    @abstractmethod
    def required_operation2(self) -> None:
        pass

    def hook1(self) -> None:
        pass

    def hook2(self) -> None:
        pass


class ConcreteClass1(AbstractClass):

    def base_operation1(self) -> None:
        print("ConcreteClass1 says: Implemented Opetation1")

    def base_operation2(self) -> None:
        print("ConcreteClass1 says: Implemented Operation2")

    def hook1(self) -> None:
        print("ConcreteClass1 says: Implemented Hook1")

    def required_operation1(self) -> None:
        print("ConcreteClass1 says: Implemented Required Operation1")

    def required_operation2(self) -> None:
        print("ConcreteClass1 says: Implemented Required Operation2")

    def hook2(self) -> None:
        print("ConcreteClass1 says: Implemented Hook2")


class ConcreteClass2(AbstractClass):

    def base_operation1(self) -> None:
        print("ConcreteClass2 says: Implemented Operation1")

    def base_operation2(self) -> None:
        print("ConcreteClass2 says: Implemented Operation2")

    def required_operation1(self) -> None:
        print("ConcreteClass2 says: Implemented Required Operation1")

    def required_operation2(self) -> None:
        print("ConcreteClass2 says: implemented Required Operation2")

    def hook1(self) -> None:
        print("ConcreteClass2 says: Implemented Hook1")

    def hook2(self) -> None:
        print("ConcreteClass2 says: Impelemented Hook2")


def client_code(abstract_class: AbstractClass):
    abstract_class.template_method()


if __name__ == "__main__":
    client_code(ConcreteClass1())
    print()
    client_code(ConcreteClass2())
